File: Day 5/part1.c
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#include "rules.h"

// Read all rules.
RuleTree *read_rules(FILE *file)
{
    RuleTree *rules = NULL;

    char *line = NULL;
    size_t buff_len = 0;
    while (getline(&line, &buff_len, file) > 1)
    {
        int before, after;
        sscanf(line, "%d|%d", &before, &after);

        RuleTree *before_node;
        rules = tree_insert(rules, before, &before_node);
        before_node->after = list_push(before_node->after, after);
    }

    free(line);

    return rules;
}

// Read a single line of page numbers.
PageList *read_pages(FILE *file)
{
    PageList *pages = NULL;

    char *line = NULL;
    size_t buff_len = 0;
    if (getline(&line, &buff_len, file) != -1)
    {
        char *page_str = strtok(line, ",");
        while (page_str != NULL)
        {
            int page_val = atoi(page_str);
            pages = list_push(pages, page_val);

            page_str = strtok(NULL, ",");
        }

        free(line);
    }

    return pages;
}

int main()
{
    FILE *file;
    file = fopen("input.txt", "r");
    if (file == NULL)
    {
        printf("Failed to open file\n");
        return EXIT_FAILURE;
    }

    RuleTree *rules = read_rules(file);

    int middle_sum = 0;
    for (PageList *pages = read_pages(file); pages != NULL; pages = read_pages(file))
    {
        bool in_order = true;
        for (PageList *curr_page = pages; curr_page != NULL; curr_page = curr_page->next)
        {
            RuleTree *rule = tree_get(rules, curr_page->value);
            if (rule == NULL)
            {
                continue;
            }

            for (PageList *after = rule->after; after != NULL; after = after->next)
            {
                if (list_contains(curr_page->next, after->value))
                {
                    in_order = false;
                    break;
                }
            }
        }

        if (in_order)
        {
            middle_sum += list_get(pages, list_len(pages) / 2);
        }

        list_free(pages);
    }

    tree_free(rules);

    printf("Middle Sum: %d\n", middle_sum);

    return EXIT_SUCCESS;
}

File: Day 14/part1.py
import re


def move(p, v, step):
    px, py = p
    vx, vy = v

    return ((px + vx*step) % 101, (py + vy*step) % 103)


def quadrant(p):
    px, py = p

    hx, hy = 101//2, 103//2

    if px < hx and py < hy:
        return 'A'
    elif px > hx and py < hy:
        return 'B'
    elif px > hx and py > hy:
        return 'C'
    elif px < hx and py > hy:
        return 'D'

    return 'Middle'


with open('input.txt') as input:
    quad_c = {}

    for line in input:
        m = re.match(r'p=(-?[0-9]*),(-?[0-9]*) v=(-?[0-9]*),(-?[0-9]*)', line)
        p = (int(m[1]), int(m[2]))
        v = (int(m[3]), int(m[4]))

        q = quadrant(move(p, v, 100))
        quad_c[q] = quad_c.get(q, 0) + 1

    safety = quad_c.get('A', 0) * quad_c.get('B', 0) * \
        quad_c.get('C', 0) * quad_c.get('D', 0)

    print("Safety factor:", safety)
